src/app/static/src/app/components/FileUpload.vue

```diff
@@ -0,0 +1,77 @@
+<template>
+    <div class="form-group">
+        <label class="font-weight-bold">{{label}}</label>
+        <tick color="#e31837" v-if="valid" width="20px"></tick>
+        <div class="custom-file">
+            <input type="file"
+                   class="custom-file-input"
+                   :ref="name"
+                   :id="name"
+                   :accept="accept"
+                   v-on:change="handleFileSelect"/>
+            <label :for="name" :class="['custom-file-label', {'selected': newFileName || existingFileName}]">
+                {{newFileName || existingFileName || "Choose a file" }}
+            </label>
+        </div>
+        <error-alert v-if="hasError" :message="error"></error-alert>
+    </div>
+</template>
+
+<script lang="ts">
+    import Vue from "vue";
+
+    import Tick from "./Tick.vue";
+    import ErrorAlert from "./ErrorAlert.vue";
+
+    interface Data {
+        newFile: File | null
+        newFileName: string
+    }
+
+    interface Computed {
+        hasError: boolean
+    }
+
+    interface Methods {
+        handleFileSelect: (_: Event, files: FileList | null) => void
+    }
+
+    type PropNames = "upload" | "accept" | "label" | "valid" | "error" | "existingFileName" | "name"
+
+    export default Vue.extend<Data, Methods, Computed, PropNames>({
+        name: "FileUpload",
+        data(): Data {
+            return {
+                newFile: null,
+                newFileName: ""
+            }
+        },
+        props: ["upload", "accept", "label", "valid", "error", "existingFileName", "name"],
+        computed: {
+            hasError: function () {
+                return this.error.length > 0
+            }
+        },
+        methods: {
+            handleFileSelect(_: Event, files: FileList | null) {
+                if (!files) {
+                    const fileInput = this.$refs[this.name] as HTMLInputElement;
+                    files = fileInput.files
+                }
+
+                this.newFile = files && files[0];
```

Comment:
Yeah this maybe slightly confusing. The newFilename is just the *selected* file, before it's been uploaded. The fileName that comes from the store is the file name after it has been uploaded (so if the page is refreshed, the store fileName will be populated.) Maybe calling the former "selectedFileName" would make more sense?

Reply:
Don't think the naming matters too much, I just wondered whether you needed to save it into the component at all, since you're sending it off round the action/mutation cycle via the upload method anyway. Which I would have expected to update the pjnzFilename in the state, which would then be picked up by the PJNZ FileUpload component. But I've just checked again and noticed that PJNZLoaded doesn't actually update that filename in the state. Which seems a little bit odd, since it has changed. 

(So how does it get the filename when you go back to the baseline step from the survey step without refreshing the page? Oh, it actually doesn't, it shows 'Choose a filename' again.)

Reply:
So the idea is that the filename displayed instantly is the selected filename, as you'd expect when selecting a file (the real upload may take a few seconds depending on bandwith/latency, maybe at some point we'll want a loading spinner or progress bar). The filename in the store is the filename that has been successfully uploaded. But yes, you're right, the PJNZUploaded should set that!